frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -1512,3 +1525,17 @@ func location(sm *llb.SourceMap, locations []parser.Range) llb.ConstraintsOpt {
 	}
 	return sm.Location(loc)
 }
+
+func summarizeHeredoc(doc string) string {
+	lines := strings.Split(strings.TrimSpace(doc), "\n")
+	first := strings.TrimSpace(lines[0])
+
+	maxLen := 32
```

Comment:
If this code is used to print progress, is it possible to make it "not truncate" when using `--progress=plain` ? I recall https://github.com/moby/buildkit/pull/1435 (also a quick look there, I see it limits at 72 chars, but not sure what's used elsewhere)

Reply:
^^ discussing with @tonistiigi what options we have. My "case" was that when using `--progress=plain` (e.g. in CI, or for debugging purposes), as a user I'd want to see the full output, in case there's an issue somewhere half-way my heredoc.

The debate is if we should do so, because heredocs (quite likely) can be _huge_ if people stuff a lot of code in them. 😓

Reply:
We don't necessarily need to trim the first line. Client-side should take care of that(hopefully it doesn't hide command that is more important than heredoc). But I don't think we should start to convert the doc to something else. Yes, you can't see the full contents and I think that is expected, shell does not print out the contents of files you execute either. User can always set `set -x` if they want to see internal commands.

Reply:
> We don't necessarily need to trim the first line.

Yes, perhaps that's an ok balance for now 👍

Reply:
We could also make sure that the line is not empty and pick the next one if it is.

Reply:
Have taken a look at the linked PR. So as I understand it, when summarizing the heredoc, select the first non-empty line, and show it as a preview - if there are more lines after, indicate more lines with `...` (or something similar)? I think that's a fairly good default in most cases.

For `--progress=plain`, I don't have strong opinions of whether heredocs should be shown in full - I'd quite like to see them in the logs, and I can't imagine they would be so much longer than the common use of `&& \` today, which aren't summarized at all. But I also get that maybe it might not be worth trying to be clever, and we could just have the same behavior throughout? I do think that either they should behave the same, or `--progress=plain` would show all the heredoc content, trying to achieve some balance in between would likely be fairly confusing.

Reply:
We had a bit of a debate about semantics; I considered "heredoc" to be somewhat equal to non-heredoc, other than different syntax in the Dockerfile, and Tonis considered them more similar to `RUN ./some-script.sh` (`some-script.sh` being the "heredoc" content).

I can see the concern about heredoc (potentially) being even longer than `&& \` being used today (as it will be easier for users to use them, so might be chosen over an actual `./script.sh`). And, agreed, for debugging, users can still use `set -x` if they want the full output (which is what they would use for an actual `./script.sh` as well).

So while I (personally) love my logs to be as verbose as possible (if I need to), I can understand the concerns, so I think we can start with printing the "first line" (e.g.). If there's a lot of demand for more verbose output, I guess we can revisit that decision in future.

Reply:
Ok, have taken a look at the logic here, how do you think the new summary looks? Caps at 72 characters, try and include multiple lines if possible, and never cut off lines half way through.

Reply:
My takeaway from this was that we should still show the first (non-empty) line only(in addition to the command). But we should not do any whitespace trimming on that line. Current changes look like the opposite.

Reply:
Ah ok, my apologies, I misunderstood - will fix that up later.

Reply:
Have taken a look, now this should just take *only* the first non-empty line (with no capping) as well as the command.

Reply:
Where is the non-empty check?

Reply:
By doing the `TrimSpace` above, that would remove all empty lines above - although it would also strip any whitespace at the beginning of the line. Is that something we want to keep as well?